gear-programs/vft-treasury/src/services/utils.rs

```diff
@@ -0,0 +1,159 @@
+    gstd::critical::set_hook(move || {
+        let msg_tracker = msg_tracker_mut();
+        let msg_info = msg_tracker
+            .get_message_info(&msg_id)
+            .expect("Unexpected: msg info does not exist");
+
+        match msg_info.status {
+            MessageStatus::SendingMessageToBridgeBuiltin => {
+                msg_tracker
+                    .update_message_status(msg_id, MessageStatus::WaitingReplyFromBuiltin)
+                    .expect("message not found");
+            }
+
+            MessageStatus::SendingMessageToTransferTokens => {
+                msg_tracker
+                    .update_message_status(msg_id, MessageStatus::WaitingReplyFromTransfer)
+                    .expect("message not found");
+            }
+
+            MessageStatus::TokenTransferCompleted(true) => {
+                msg_tracker
+                    .update_message_status(msg_id, MessageStatus::BridgeBuiltinStep)
+                    .expect("message not found");
+            }
+
+            MessageStatus::TokenTransferCompleted(false) => {
+                msg_tracker.remove_message_info(&msg_id);
+            }
+
+            MessageStatus::BridgeResponseReceived(None) => {}
+            _ => {}
+        }
+    });
+}
+
+fn decode_transfer_reply(bytes: &[u8]) -> Result<bool, Error> {
+    vft_io::TransferFrom::decode_reply(bytes).map_err(|_| Error::TransferTokensDecode)
+}
+
+fn decode_bridge_reply(mut bytes: &[u8]) -> Result<Option<U256>, Error> {
+    let reply =
+        gbuiltin_eth_bridge::Response::decode(&mut bytes).map_err(|_| Error::BuiltinDecode)?;
+
+    match reply {
+        gbuiltin_eth_bridge::Response::EthMessageQueued { nonce, .. } => Ok(Some(nonce)),
+    }
+}
+
+fn handle_reply_hook(msg_id: MessageId) {
+    let msg_tracker = msg_tracker_mut();
+
+    let msg_info = msg_tracker
+        .get_message_info(&msg_id)
+        .expect("Unexpected: msg info does not exist");
+    let reply_bytes = gstd::msg::load_bytes().expect("Unable to load bytes");
+
+    match msg_info.status {
+        MessageStatus::SendingMessageToTransferTokens => {
+            match decode_transfer_reply(&reply_bytes) {
+                Ok(reply) => {
+                    msg_tracker
+                        .update_message_status(msg_id, MessageStatus::TokenTransferCompleted(reply))
+                        .expect("message not found");
+                }
+
+                Err(_) => {
+                    msg_tracker.remove_message_info(&msg_id);
+                }
+            }
+        }
+
+        MessageStatus::WaitingReplyFromTransfer => {
+            let reply = decode_transfer_reply(&reply_bytes).unwrap_or(false);
+
+            if reply {
+                msg_tracker
+                    .update_message_status(msg_id, MessageStatus::BridgeBuiltinStep)
+                    .expect("message not found");
+            } else {
+                msg_tracker.remove_message_info(&msg_id);
+            }
+        }
```

Comment:
Do we really need a `BridgeBuiltinStep` status, it seems duplicate of `MessageStatus::TokenTransferCompleted(true)`?